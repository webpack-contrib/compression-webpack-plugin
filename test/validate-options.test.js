import CompressionPlugin from '../src';

import { removeCache } from './helpers';

describe('validate options', () => {
  beforeEach(() => {
    return removeCache();
  });

  const tests = {
    test: {
      success: [
        /foo/,
        'foo',
        [/foo/],
        [/foo/, /bar/],
        ['foo', 'bar'],
        [/foo/, 'bar'],
      ],
      failure: [true, [true], [/foo/, 'foo', true]],
    },
    include: {
      success: [
        /foo/,
        'foo',
        [/foo/],
        [/foo/, /bar/],
        ['foo', 'bar'],
        [/foo/, 'bar'],
      ],
      failure: [true, [true], [/foo/, 'foo', true]],
    },
    exclude: {
      success: [
        /foo/,
        'foo',
        [/foo/],
        [/foo/, /bar/],
        ['foo', 'bar'],
        [/foo/, 'bar'],
      ],
      failure: [true, [true], [/foo/, 'foo', true]],
    },
    filename: {
      success: ['[path].gz[query]', () => {}],
      failure: [true],
    },
    algorithm: {
      success: ['gzip', () => {}],
      failure: [true],
    },
    compressionOptions: {
      success: [{ level: 1 }, { unknown: 1 }],
      failure: ['1024'],
    },
    threshold: {
      success: [1024],
      failure: ['1024'],
    },
    minRatio: {
      success: [0.8],
      failure: ['0.8'],
    },
    cache: {
      success: [true, '/path/to/cache'],
      failure: [() => {}],
    },
    deleteOriginalAssets: {
      success: [true, false],
      failure: ['true'],
    },
    unknown: {
      success: [],
      failure: [1, true, false, 'test', /test/, [], {}, { foo: 'bar' }],
    },
  };

  function stringifyValue(value) {
    if (
      Array.isArray(value) ||
      (value && typeof value === 'object' && value.constructor === Object)
    ) {
      return JSON.stringify(value);
    }

    return value;
  }

  async function createTestCase(key, value, type) {
    it(`should ${
      type === 'success' ? 'successfully validate' : 'throw an error on'
    } the "${key}" option with "${stringifyValue(value)}" value`, async () => {
      let error;

      try {
        // eslint-disable-next-line no-new
        new CompressionPlugin({ [key]: value });
      } catch (errorFromPlugin) {
        if (errorFromPlugin.name !== 'ValidationError') {
          throw errorFromPlugin;
        }

        error = errorFromPlugin;
      } finally {
        if (type === 'success') {
          expect(error).toBeUndefined();
        } else if (type === 'failure') {
          expect(() => {
            throw error;
          }).toThrowErrorMatchingSnapshot();
        }
      }
    });
  }

  for (const [key, values] of Object.entries(tests)) {
    for (const type of Object.keys(values)) {
      for (const value of values[type]) {
        createTestCase(key, value, type);
      }
    }
  }
});
