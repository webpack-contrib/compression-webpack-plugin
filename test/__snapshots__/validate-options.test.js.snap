// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validation 1`] = `
"Invalid configuration object. Object has been initialised using a configuration object that does not match the API schema.
 - configuration.test should be one of these:
   RegExp | string | [RegExp | string, ...]
   Details:
    * configuration.test should be one of these:
      RegExp | string
      Details:
       * configuration.test should be an instance of RegExp.
       * configuration.test should be a string.
    * configuration.test should be an array:
      [RegExp | string, ...]"
`;

exports[`validation 2`] = `
"Invalid configuration object. Object has been initialised using a configuration object that does not match the API schema.
 - configuration.test should be one of these:
   RegExp | string | [RegExp | string, ...]
   Details:
    * configuration.test[0] should be one of these:
      RegExp | string
      Details:
       * configuration.test[0] should be an instance of RegExp.
       * configuration.test[0] should be a string."
`;

exports[`validation 3`] = `
"Invalid configuration object. Object has been initialised using a configuration object that does not match the API schema.
 - configuration.test should be one of these:
   RegExp | string | [RegExp | string, ...]
   Details:
    * configuration.test[2] should be one of these:
      RegExp | string
      Details:
       * configuration.test[2] should be an instance of RegExp.
       * configuration.test[2] should be a string."
`;

exports[`validation 4`] = `
"Invalid configuration object. Object has been initialised using a configuration object that does not match the API schema.
 - configuration.include should be one of these:
   RegExp | string | [RegExp | string, ...]
   Details:
    * configuration.include should be one of these:
      RegExp | string
      Details:
       * configuration.include should be an instance of RegExp.
       * configuration.include should be a string.
    * configuration.include should be an array:
      [RegExp | string, ...]"
`;

exports[`validation 5`] = `
"Invalid configuration object. Object has been initialised using a configuration object that does not match the API schema.
 - configuration.include should be one of these:
   RegExp | string | [RegExp | string, ...]
   Details:
    * configuration.include[0] should be one of these:
      RegExp | string
      Details:
       * configuration.include[0] should be an instance of RegExp.
       * configuration.include[0] should be a string."
`;

exports[`validation 6`] = `
"Invalid configuration object. Object has been initialised using a configuration object that does not match the API schema.
 - configuration.include should be one of these:
   RegExp | string | [RegExp | string, ...]
   Details:
    * configuration.include[2] should be one of these:
      RegExp | string
      Details:
       * configuration.include[2] should be an instance of RegExp.
       * configuration.include[2] should be a string."
`;

exports[`validation 7`] = `
"Invalid configuration object. Object has been initialised using a configuration object that does not match the API schema.
 - configuration.exclude should be one of these:
   RegExp | string | [RegExp | string, ...]
   Details:
    * configuration.exclude should be one of these:
      RegExp | string
      Details:
       * configuration.exclude should be an instance of RegExp.
       * configuration.exclude should be a string.
    * configuration.exclude should be an array:
      [RegExp | string, ...]"
`;

exports[`validation 8`] = `
"Invalid configuration object. Object has been initialised using a configuration object that does not match the API schema.
 - configuration.exclude should be one of these:
   RegExp | string | [RegExp | string, ...]
   Details:
    * configuration.exclude[0] should be one of these:
      RegExp | string
      Details:
       * configuration.exclude[0] should be an instance of RegExp.
       * configuration.exclude[0] should be a string."
`;

exports[`validation 9`] = `
"Invalid configuration object. Object has been initialised using a configuration object that does not match the API schema.
 - configuration.exclude should be one of these:
   RegExp | string | [RegExp | string, ...]
   Details:
    * configuration.exclude[2] should be one of these:
      RegExp | string
      Details:
       * configuration.exclude[2] should be an instance of RegExp.
       * configuration.exclude[2] should be a string."
`;

exports[`validation 10`] = `
"Invalid configuration object. Object has been initialised using a configuration object that does not match the API schema.
 - configuration.cache should be one of these:
   boolean | string
   Details:
    * configuration.cache should be a boolean.
    * configuration.cache should be a string."
`;

exports[`validation 11`] = `
"Invalid configuration object. Object has been initialised using a configuration object that does not match the API schema.
 - configuration.filename should be one of these:
   string | function
   Details:
    * configuration.filename should be a string.
    * configuration.filename should be an instance of function."
`;

exports[`validation 12`] = `
"Invalid configuration object. Object has been initialised using a configuration object that does not match the API schema.
 - configuration.algorithm should be one of these:
   string | function
   Details:
    * configuration.algorithm should be a string.
    * configuration.algorithm should be an instance of function."
`;

exports[`validation 13`] = `
"Invalid configuration object. Object has been initialised using a configuration object that does not match the API schema.
 - configuration.compressionOptions should be an object:
   object { â€¦ }"
`;

exports[`validation 14`] = `
"Invalid configuration object. Object has been initialised using a configuration object that does not match the API schema.
 - configuration.threshold should be a number."
`;

exports[`validation 15`] = `
"Invalid configuration object. Object has been initialised using a configuration object that does not match the API schema.
 - configuration.minRatio should be a number."
`;

exports[`validation 16`] = `
"Invalid configuration object. Object has been initialised using a configuration object that does not match the API schema.
 - configuration.deleteOriginalAssets should be a boolean."
`;

exports[`validation 17`] = `
"Invalid configuration object. Object has been initialised using a configuration object that does not match the API schema.
 - configuration has an unknown property 'unknown'. These properties are valid:
   object { test?, include?, exclude?, cache?, filename?, algorithm?, compressionOptions?, threshold?, minRatio?, deleteOriginalAssets? }"
`;
